File: examples/psi/java/EcdhPsi.java
// Copyright 2024 Ant Group Co., Ltd. #
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import com.github.fmeum.rules_jni.RulesJni;
import java.lang.annotation.Native;

public class EcdhPsi {
  static {
    RulesJni.loadLibrary("ecdh_psi_jni", EcdhPsi.class);
  }

  // Constructor
  EcdhPsi() {
    ec_key_ = jni_ecc_keygen();
    inited_ = true;
  }

  public byte[][] MaskStrings(String[] in) {
    if (!this.inited_) {
      throw new IllegalArgumentException("EcdhPsi instance has not been inited");
    }
    return jni_mask_strings(in, this.GetEcKey());
  }

  public byte[][] MaskEcPointAndHashTo128(byte[][] in) {
    if (!this.inited_) {
      throw new IllegalArgumentException("EcdhPsi instance has not been inited");
    }
    return jni_mask_ec_point_and_hash_to_u128(in, this.GetEcKey());
  }

  byte[] GetEcKey() {
    return ec_key_;
  }

  private byte[] ec_key_;
  private boolean inited_ = false;

  // ----------------------------
  // Native Functions Declaration
  // ----------------------------
  //
  // NOTE the gerenated c header (*.h) file could be found at
  // bazel-bin/examples/psi/java/EcdhPsi.hdrs.h/EcdhPsi.h

  public static native byte[] jni_ecc_keygen();

  public static native byte[][] jni_mask_strings(String[] in, byte[] ec_key);

  public static native byte[][] jni_mask_ec_point_and_hash_to_u128(byte[][] in, byte[] ec_key);
}

File: examples/psi/cpp/ecdh_psi.cc
// Copyright 2023 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "psi/cpp/ecdh_psi.h"

#include <memory>
#include <vector>

#include "yacl/crypto/ecc/ec_point.h"
#include "yacl/crypto/ecc/ecc_spi.h"
#include "yacl/crypto/hash/hash_utils.h"
#include "yacl/link/link.h"
#include "yacl/secparam.h"

namespace yc = yacl::crypto;

namespace examples::psi {

void EcdhPsi::MaskStrings(absl::Span<std::string> in,
                          absl::Span<yc::EcPoint> out) const {
  YACL_ENFORCE(!in.empty());
  YACL_ENFORCE(in.size() == out.size());
  for (size_t i = 0; i < in.size(); ++i) {
    out[i] = ec_->HashToCurve(yc::HashToCurveStrategy::Autonomous, in[i]);
    ec_->MulInplace(&out[i], sk_);
  }
}

void EcdhPsi::MaskEcPointsAndHashToU128(absl::Span<yc::EcPoint> in,
                                        absl::Span<uint128_t> out) const {
  YACL_ENFORCE(!in.empty());
  YACL_ENFORCE(in.size() == out.size());
  for (size_t i = 0; i < in.size(); ++i) {
    out[i] = yc::Blake3_128(ec_->SerializePoint(ec_->Mul(in[i], sk_)));
  }
}

// Mask input strings with secret key, and outputs the EcPoint results
std::vector<std::string> EcdhPsi::MaskStringsEx(
    std::vector<std::string> in) const {
  YACL_ENFORCE(!in.empty());
  std::vector<std::string> out(in.size());
  for (size_t i = 0; i < in.size(); ++i) {
    auto temp = ec_->HashToCurve(yc::HashToCurveStrategy::Autonomous, in[i]);
    ec_->MulInplace(&temp, sk_);
    out[i] = ec_->SerializePoint(temp);
  }
  return out;
}

// Mask input strings with secret key, and outputs the EcPoint results
std::vector<uint128_t> EcdhPsi::MaskEcPointsAndHashToU128Ex(
    std::vector<std::string> in) const {
  YACL_ENFORCE(!in.empty());
  std::vector<uint128_t> out(in.size());
  for (size_t i = 0; i < in.size(); ++i) {
    auto temp = ec_->DeserializePoint(in[i]);
    out[i] = yc::Blake3_128(ec_->SerializePoint(ec_->Mul(temp, sk_)));
  }
  return out;
}
}  // namespace examples::psi
